### ISSUE TITLE
"ProductDetails" method is invoked when not-existing some static file in the root is requested

### ISSUE THREAD

**Issue by AndreiMaz** [nopSolutions/nopCommerce#2922]
Originally the issue was described at https://github.com/nopSolutions/nopCommerce/issues/2902. It was fixed.

But for static files in the root (e.g. www.site.com/picture.jpg) it invokes the "ProductDetails" method. That's wrong. It should be ignored

**Comment by davidandrewpowell** [comment]
I believe that the reason that this issue is occurring is due to how outgoing generic URLs are generated in Nop.Web/Infrastructure/GenericUrlRouteProvider.cs for products, categories, manufacturers, vendors, news items, blog posts, topics and products by tag.

I believe that requested content (in our example `/picture.jpg`) will first be searched for among the static files (using the setup made in `application.UseStaticFiles()`).  When the image is not found it then proceeds to try to match the requested URL against the routes that are registered by the application in `RouteProvider `and `GenericUrlRouteProvider`.

First it goes through all of the routes in `RouteProvider.cs` (as these were registered first).  There is no match in here for `/picture.jpg`.

So then it attempts to match the requested URL against the routes in `GenericUrlRouteProvider.cs`:

The first route to match against in `GenericUrlRouteProvider` is the default route:

`routeBuilder.MapRoute("Default", "{controller}/{action}/{id?}");`

This route **does not match** the requested URL.

The second route to match against in `GenericUrlRouteProvider` is the GenericUrl route designed to match against single segment URLs such as `/desktops`:

```
routeBuilder.MapGenericPathRoute("GenericUrl", "{GenericSeName}", 
    new { controller = "Common", action = "GenericUrl" });
```

Our `/picture.jpg` is also a single segment URL.  This route matches the slug 'picture.jpg' against the route variable '{GenericSeName}' and then attempts to search for it in the database.

This route **does not match** the requested URL.


The third route to match against is the Product route:

```
routeBuilder.MapLocalizedRoute("Product", "{SeName}", 
    new { controller = "Product", action = "ProductDetails" });
```

This route consists of a single segment "{SeName}" and so **this does match** the requested URL of `/picture.jpg`.  **This is the issue that is occurring.**


To resolve this issue, this will involve removing the following routes that will all match the single segment:

```
routeBuilder.MapLocalizedRoute("Product", "{SeName}", new { controller = "Product", action = "ProductDetails" });

routeBuilder.MapLocalizedRoute("Category", "{SeName}", 
    new { controller = "Catalog", action = "Category" });

routeBuilder.MapLocalizedRoute("Manufacturer", "{SeName}", 
    new { controller = "Catalog", action = "Manufacturer" });

routeBuilder.MapLocalizedRoute("Vendor", "{SeName}", 
    new { controller = "Catalog", action = "Vendor" });
            
routeBuilder.MapLocalizedRoute("NewsItem", "{SeName}", 
    new { controller = "News", action = "NewsItem" });

routeBuilder.MapLocalizedRoute("BlogPost", "{SeName}", 
    new { controller = "Blog", action = "BlogPost" });

routeBuilder.MapLocalizedRoute("Topic", "{SeName}", 
    new { controller = "Topic", action = "TopicDetails" });

routeBuilder.MapLocalizedRoute("ProductsByTag", "{SeName}",
    new { controller = "Catalog", action = "ProductsByTag" });
```

It is correct that they should be removed though as they are essentially all duplicates of the 'GenericUrl' route but with different names.

Removing these routes will fix the issue but will generate a new issue as outgoing URLs for these routes will no longer be generated.

For example, in the Topic.cshtml you will find the following code to generate a generic URL:

`@Url.RouteUrl("Topic", new {SeName = Model.SeName})`

Here, the 'Topic' route is referenced specifically in order to generate the URL.

Instead, this can be changed to:

`@Url.RouteUrl("GenericUrl", new {SeName = Model.SeName})`

and this will still generate the desired URL.

**Comment by davidandrewpowell** [comment]
I have added a [pull request](https://github.com/nopSolutions/nopCommerce/pull/3721) to fix this issue using the fix described above.

**Comment by davidandrewpowell** [comment]
I thought I should mention also, that this issue is not just for static files, it is for any single segment route.  So for example, both of the following routes will invoke the ProductDetails action:

[www.site.com/nonexistantpicture.jpg]()
[www.site.com/nomatchingroute]()

These are both fixed in the suggestion and PR mentioned above.

**Comment by davidandrewpowell** [comment]
@RomanovM  I noticed that you had added the 'on hold / maybe wont' tag to this issue.

I was a little concerned that this PR may not get merged, possibly due to the large number of places that need to call the '_GenericUrl_' route so I just wanted to run this by you.

In a nutshell the '_GenericUrl_' route AND the '_Product_', '_Category_', '_Manufacturer_', '_Vendor_', '_NewsItem_', '_BlogPost_', '_Topic_' and '_ProductsByTag_' routes all have the same route format of "`{sename}`".  The only difference is that '_GenericUrl_' is a `GenericPathRoute` and the others are `LocalizedRoute`'s.

For urls such as [www.site.com/doesnotexist](), it tries to match the '_doesnotexist_' part against the `GenericUrl` route (looking at the slugs in the database).  When these don't match it then tries to match this part against the next route which is the '_Product_' route.  This matches the '`{sename}`' segment and thus calls the `ProductDetails` action.

The '_Product_', '_Category_', '_Manufacturer_', '_Vendor_', '_NewsItem_', '_BlogPost_', '_Topic_' and '_ProductsByTag_' routes are all duplicates of '_GenericUrl_' (by pattern) and are not required to match **incoming** urls or create **outgoing** urls as the '_GenericUrl_' does this fine.

I have painstakingly gone through the code to find all references to these old routes as much as possible.  As (I'm sure you are aware) there may be a very small chance that I might have missed one, so I would not suggest merging this PR into the 4.2 release at this stage.  But if possible, I would suggest merging it in to the 4.3 develop branch after 4.2 has been released.  Hopefully then if there are any missed route names, these will reveal themselves.  I am confident that I have found all that I can.

Whilst none of this is ideal, I do believe that the routing should be corrected around the url slug matching functionality.

I will leave this with your best judgement.

Thanks
